p2p/src/sync/mod.rs

```diff
@@ -73,182 +127,2369 @@ where
+            }
+
+            self.state = SyncState::DownloadingBlocks;
+            return Ok(());
+        }
+
+        // try to drain the queue if there are any blocks
+        for chain in self.queue.drain().iter() {
+            for block in chain.iter() {
+                self.p2p_handle.new_block((**block).clone()).await?;
             }
         }
 
+        // TODO: refresh all peer states here?
+        self.state = SyncState::Idle;
         Ok(())
     }
 
-    /// Run SyncManager event loop
+    /// Run [SyncManager] event loop
     pub async fn run(&mut self) -> error::Result<()> {
         log::info!("starting sync manager event loop");
 
         loop {
             tokio::select! {
-                res = self.handle.poll_next() => {
-                    self.on_floodsub_event(res?)?;
-                }
-                res = self.rx_sync.recv().fuse() => {
-                    self.on_sync_event(res.ok_or(P2pError::ChannelClosed)?).await?;
-                }
-                res = self.rx_peer.recv().fuse() => {
-                    self.on_peer_event(res.ok_or(P2pError::ChannelClosed)?).await?;
+                res = self.rx_floodsub.recv().fuse() => {
+                    let event::SyncFloodEvent::Block { peer_id, block } = res.ok_or(P2pError::ChannelClosed)?;
+                    self.process_block(peer_id, Arc::new(block)).await.into_fatal()?;
+                },
+                event = self.rx_sync.recv().fuse() => match event.ok_or(P2pError::ChannelClosed)? {
+                    event::SyncControlEvent::Connected { peer_id, tx } => {
+                        self.register_peer(peer_id, tx).await.into_fatal()?;
+                    }
+                    event::SyncControlEvent::Disconnected { peer_id } => {
+                        self.unregister_peer(peer_id).await.into_fatal()?;
+                    }
+                },
+                event = self.rx_peer.recv().fuse() => match event.ok_or(P2pError::ChannelClosed)? {
+                    event::PeerSyncEvent::Headers { peer_id, headers } => {
+                        self.initialize_peer(peer_id, &headers).await.into_fatal()?;
+                    }
+                    event::PeerSyncEvent::GetHeaders { peer_id, locator } => {
+                        self.process_header_request(peer_id, locator).await.into_fatal()?;
+                    }
+                    event::PeerSyncEvent::Blocks { peer_id, blocks } => {
+                        self.process_block_response(peer_id, blocks).await.into_fatal()?;
+                    }
+                    event::PeerSyncEvent::GetBlocks { peer_id, headers } => {
+                        self.process_block_request(peer_id, &headers).await.into_fatal()?;
+                    }
                 }
             }
+
+            // advance the state of the manager after each event has been handled
+            self.advance_state().await.into_fatal()?;
         }
     }
 }
 
 #[cfg(test)]
```

Comment:
Huge tests... maybe move them to their own file?

Reply:
Yeah some of them don't fall into the unit testing category so I'll see if I can move them somewhere else.